ocaml-lsp-server/src/semantic_highlighting.ml

```diff
@@ -0,0 +1,828 @@
+
+  let module_type (self : Ast_iterator.iterator) (mt : Parsetree.module_type) =
+    match
+      match mt.pmty_desc with
+      | Pmty_ident l ->
+        lident l Module_type ();
+        `Custom_iterator
+      | Pmty_functor (fp, mt) ->
+        (match fp with
+        | Unit -> ()
+        | Named (n, mt) ->
+          add_token n.loc Module [];
+          self.module_type self mt);
+        self.module_type self mt;
+        `Custom_iterator
+      | Pmty_alias m ->
+        lident m Module ();
+        `Custom_iterator
+      | Pmty_signature sis ->
+        List.iter sis ~f:(fun si -> self.signature_item self si);
+        `Custom_iterator
+      | Pmty_with (_, _) | Pmty_typeof _ | Pmty_extension _ -> `Default_iterator
+    with
+    | `Custom_iterator -> ()
+    | `Default_iterator -> Ast_iterator.default_iterator.module_type self mt
+
+  let signature_item (self : Ast_iterator.iterator)
+      (si : Parsetree.signature_item) =
+    match
+      (* TODO *)
+      match si.psig_desc with
+      | Psig_typext _ | Psig_typesubst _
+      | Psig_type (_, _)
+      | Psig_value _
+      | Psig_modtype _
+      | Psig_exception _
+      | Psig_module _
+      | Psig_modsubst _
+      | Psig_recmodule _
+      | Psig_modtypesubst _
+      | Psig_open _
+      | Psig_include _
+      | Psig_class _
+      | Psig_class_type _
+      | Psig_attribute _
+      | Psig_extension (_, _) -> `Default_iterator
+    with
+    | `Default_iterator -> Ast_iterator.default_iterator.signature_item self si
+    | `Custom_iterator -> ()
+
+  (* TODO: *)
+  let attribute _self _attr = ()
+
+  (* TODO: *)
+  let attributes _self _attrs = ()
+
+  let iterator =
+    { Ast_iterator.default_iterator with
+      module_binding
+    ; type_declaration
+    ; expr
+    ; pat
+    ; constructor_declaration
+    ; label_declaration
+    ; typ
+    ; value_binding
+    ; signature_item
+    ; module_type_declaration
+    ; attribute
+    ; attributes
+    ; module_expr
+    ; value_description
+    ; module_type
+    ; module_declaration
+    }
+
+  let apply parsetree =
+    let () =
+      match parsetree with
+      | `Interface signature -> iterator.signature iterator signature
+      | `Implementation structure -> iterator.structure iterator structure
+    in
+    Tokens.encode_tokens tokens
+end
+
+(* Note: this is global mutable state; we could move this to [State.t] *)
+let gen_new_id =
+  let i = ref 0 in
+  fun () ->
+    let x = !i in
+    incr i;
+    string_of_int x
+
+let on_request_full :
+    State.t -> SemanticTokensParams.t -> SemanticTokens.t option Fiber.t =
+ fun state params ->
+  Fiber.of_thunk (fun () ->
+      let store = state.store in
+      let uri = params.textDocument.uri in
+      let doc = Document_store.get store uri in
+      let+ tokens =
+        Document.with_pipeline_exn doc (fun pipeline ->
+            let parsetree = Mpipeline.reader_parsetree pipeline in
+            let module Fold = Parsetree_fold () in
```

Comment:
You don't want to do unnecessary work in the merlin thread. The fold should be safe outside as the parsetree is immutable.

Reply:
Right, forgot to fix this after prototyping. Fixed now.